refactored mnemonic

jimmysong opened the pull request:
took out asserts
put words in a separate file
lazy initialization
functions to do word lookups instead of list/dict

mflaxman commented on buidl/mnemonic.py:
```diff
@@ -34,2082 +38,63 @@ def secure_mnemonic(num_bits=256, extra_entropy=0):
+def __init__():
+    # load word from file and make a lookup table
+    word_file = path.join(path.dirname(__file__), "bip39_words.txt")
+    global BIP39_WORDS, BIP39_LOOKUP
```
Is a global needed? Can't this code block just run every time this file is loaded (outside the `_init_()`.

I don't have a lot of experience with globals and thought the best practice was to avoid them unless there was a strong reason. I could be wrong though.

mflaxman commented on buidl/mnemonic.py:
```diff
@@ -34,2082 +38,63 @@ def secure_mnemonic(num_bits=256, extra_entropy=0):
+    if not BIP39_WORDS:
+        __init__()
+    for word in BIP39_WORDS:
+        yield word
```
Is there a performance impact here? I would think the performance bottleneck would be reading the `txt` file (which happens all at once), not iterating through the words.

Not a big deal.

mflaxman reviewed (approved):
I like the new error raising (not asserts), but am not sure I understand the purpose of the rest. it seems to be more complex now and I'd think this would be trickier for a new developer to intuit what's going on.

Another way to improve readability of the block of bip39_words would be to make a file called something like `bip39_wordlist.py` that had only one entry:
```
BIP39_WORDS = """
abandon
ability
able
...
zone
zoo
""".strip().split()
```

I don't feel strongly, feel free to merge it however you like. I appreciate the constant iteration!

mflaxman commented:
Just noticed the branch name, are you adding SLIP39 support? Neat!

jimmysong commented:
I'm working on it, yes. Need it for some backups as you so eloquently argued for =)

jimmysong commented:
OK, refactored this and added basic SLIP39 support. Only supports one level, not two like the spec, but that seems like overkill to me. Multisig should be the first level and SLIP39 the second. I suppose a third level could be used there as well, but I'm not sure what sort of use case that would be.

BIP39 and SLIP39 are classes that inherit from each other to make the code DRY. I probably need to add more tests, but this is a good start.

mflaxman commented on buidl/mnemonic.py:
```diff
@@ -26,2090 +30,86 @@ def secure_mnemonic(num_bits=256, extra_entropy=0):
-        if len(word) > 4:
-            WORD_LOOKUP[word[:4]] = i
+    @classmethod
+    def word(cls, index):
```
Maybe `get_word(cls, index):` would be less ambiguous?

jimmysong replied:
I'll do `__getitem__` so it works more like a dict.

mflaxman commented on buidl/mnemonic.py:
```diff
@@ -26,2090 +30,86 @@ def secure_mnemonic(num_bits=256, extra_entropy=0):
+                cls.lookup[word[:4]] = i
+
+    @classmethod
+    def index(cls, word):
```
Maybe `get_index()` would be less ambiguous? On second thought, `get_index()` could imply that if there wasn't a matching word you return `None` instead of raising an exception.

I think the correct way to call this might just be `WordList.lookup[word]` (get rid of the helper method all together) instead of `WordList.index(word)`. This is why I find classes tricky, because the caller has to look deep under the hood to figure out how they work. The previous `BIP39_WORD_LIST[word]` seems strictly superior to the new class-based structure.

mflaxman replied:
Also, should this lookup be for `word.lower()` to be more forgiving?

I was thinking that would happen as part of the `normalize()` function, but in the current code-path this method is called first from `from_mnemonic()`.

jimmysong replied:
I was thinking `.index` is a method available to lists. Maybe I should subclass a data structure. Let me see if I can make it more usable that way.

mflaxman commented on buidl/shamir.py:
```diff
@@ -0,0 +1,329 @@
+                "sha256",
+                i + passphrase,
+                self.salt + right,
+                2500 << self.exponent,  # 10000 << self.exponent // 4
```
Delete commented out code

jimmysong replied:
this is actually a comment to show that the formula is the same =)

mflaxman commented on buidl/shamir.py:
```diff
@@ -0,0 +1,329 @@
+                "sha256",
+                i + passphrase,
+                salt + right,
+                2500 << exponent,  # 10000 << exponent // 4
```
Delete commented out code

jimmysong replied:
same as above, though now that I look at it, maybe I can combine them.

mflaxman commented on buidl/mnemonic.py:
```diff
@@ -26,2090 +30,86 @@ def secure_mnemonic(num_bits=256, extra_entropy=0):
+    def _load(cls):
+        word_file = path.join(path.dirname(__file__), cls.filename)
+        with open(word_file, "r") as f:
+            cls.words = f.read().split()
```
This should throw an exception/crash if for some reason it is unable to parse this file. I can't think of why that would happen, perhaps on some OS the file path works differently and the text file isn't there.

jimmysong replied:
SLIP39 is 1024 words, unfortunately, though I can put expected number or something in there.

mflaxman commented on buidl/test/test_hd.py:
```diff
@@ -404,7 +404,8 @@ def test_bech32_address(self):
         for word in mnemonic.split():
-            self.assertTrue(word in WORD_LIST)
+            # should throw an error if this is not a word
+            BIP39.index(word)
```
Repeated point from before, but using a method like this makes it non-obvious that it should throw an error if this is not a word, whereas a normal dictionary lookup doesn't have this ambiguity.

jimmysong replied:
yea, i'll use `__getitem__` so it works like a dictionary lookup.

mflaxman commented on buidl/hd.py:
```diff
@@ -377,7 +377,7 @@ def from_mnemonic(cls, mnemonic, password=b"", path="m", testnet=False):
         normalized_words = []
         for word in words:
-            normalized_words.append(WORD_LIST[WORD_LOOKUP[word]])
+            normalized_words.append(BIP39.normalize(word))
```
This comment predates this PR, but I noticed we are doing `for word in words:` 2x here, once above for the index and once down here to normalize the words. I think this loop should be done 1x in the same place.

mflaxman reviewed (approved):
Awesome to have slip39 support! I didn't audit the math behind that closely, just commenting on the structure/features.

Overall, I'm not in love with the new class-based structure for BIP39 (and SLIP39) word lists. I think the added ambiguity/complexity is a negative for application developers who don't have your depth of understanding. That said, I don't think it's the end of the world and defer to you. I love your pace of iteration and don't want to slow that down with bikeshedding.

Is the reason for not adding all the test vectors [example](https://github.com/trezor/python-shamir-mnemonic/blob/master/vectors.json#L24-L30) because of the 1-level? I agree the use-case for multi-level is very limited, but cross-compatibility with other implementations (i.e. Trezor) seems essential. I'd like to add this to multiwallet and have it work out of the box with Hermit for example (I have a PSBT branch with QR support that I'm working on). For the latter, buidl can only be used if it supports multi-level so that it can replace what's currently there.

As before, please merge when you see fit!

jimmysong commented:
The code will now be able to recover 2-layer shamir. It will not generate it, though. Useful for recovery.

Now every test case except the last one is in the unit test.

jimmysong commented:
@mflaxman please take a look one more time to see if your concerns have been addressed.

mflaxman commented on buidl/test/test_shamir.py:
```diff
@@ -0,0 +1,370 @@
+        ]
+        for test_name, mnemonics, error in test_cases:
+            with self.assertRaises(error):
+                print(test_name)
```
Print statement

mflaxman commented on buidl/hd.py:
```diff
@@ -355,11 +355,15 @@ def from_mnemonic(cls, mnemonic, password=b"", path="m", testnet=False):
         # each word is 11 bits
+        normalized_words = []
         for word in words:
             # get the number that the word represents using WORD_LOOKUP
```
Outdated variable name in comment

mflaxman reviewed (approved):
Have still not audited the math, but code LGTM.

mflaxman commented:
Wohoo! @jimmysong, want to add this to multiwallet.py?
